### PR TITLE
Add ability to break parking lots, stop locks from stalling

### LINKED ISSUE #3035: Problems with `trio.Lock()` across multiple tasks
Some time ago - never mind how long precisely - having a little coordination between tasks and nothing particular to interest me in semaphores, I saw our `RuntimeError("can't release a Lock you don't own")`.

I approve pretty strongly of keeping things local by default; it prevents all sorts of bugs and `trio.Semaphore` is right there if you need something without this limitation.  But two further observations prompted this issue:

1. `trio.Semaphore()` is considerably more powerful and flexible in general.  By the principle of least power, it might be nice to have `trio.Lock(*, allow_outside_release: bool = False)` or something along those lines.
2. If the owning task finishes (returns or raises) _without releasing the lock_, then you're stuck forever.  The toy program below demonstrates.  Perhaps we should raise an error or at least emit a warning when a task which owns a lock finishes?

```python
import trio

@trio.run
async def main():
    lock = trio.Lock()
    async with trio.open_nursery() as nursery:
        nursery.start_soon(lock.acquire)
    # The lock is now held, and can only be released by the acquiring task.
    # However, that task is now finished, so we can never release the lock!
    lock.release()
```

### PR DESCRIPTION
**Pull request by jakkdl**
fixes #3035 

TODO:
* [x] newsfragment(s)
* [x] docstrings
* [x] docs
* [x] tests of directly using the parking lot functions
* [x] lock test that checks that handing over the owner works (this is indirectly tested by a lot of other tests, but good to have a standalone test)
* [x] reraise the exception in `Lock`/`_LockImpl` to get a better error message

### CODE REVIEW

File: src/trio/_core/_parking_lot.py
Comment (TeamSpen210) [inline src/trio/_core/_parking_lot.py:109]: Instead of directly checking each time, this could just catch the `KeyError` (then reraise and suppress), so it has zero overhead in the happy case. Might also want to catch `ValueError` from `remove()` failing.

File: src/trio/_core/_parking_lot.py
Comment (TeamSpen210) [inline src/trio/_core/_parking_lot.py:277]: I think this shouldn't be cached, each task should get its own outcome & exception object. Otherwise the traceback would get scrambled together.

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py:109]: Fixed

and tbh it's probably better not to distinguish between `KeyError` and `ValueError`. Which one you'll get will depend on if the task has previously parked any lots or not, not much about the current state.

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py:277]: thanks, fixed

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:96]: Hm, this isn't exactly the behavior we're looking for: it's if it exits without removing itself as a parking lot breaker, rather than if it exits while parked on the lot.

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:280]: I agree that it should be reraised -- I think it's best to use a subclass so that anyone using this can catch and reraise without any concern about catching things incorrectly + still have access to the task that caused the parking lot to break.

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:272]: I think taking `task` optionally might be a nicer interface (cause we're making this public I guess), defaulting it to `trio.lowlevel.current_task()`.

**Comment by A5rocks** [comment]
@njsmith cause this is interface design, do you think there's a better way to express all of this?

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:278]: Probably some thin veneer of thread safety? Or guaranteeing that we'll remove every task we raise an error in, even if somehow `_core.reschedule` raises an error.

Neither is really something we care about but probably worth using cause it's a simple drop in replacement anyways.

Comment (TeamSpen210) [reply src/trio/_core/_parking_lot.py:278]: `unpark()` above would have the same theoretical issue, if `reschedule()` failed it could drop tasks too. Probably not something to worry about, if that fails we've got bigger problems.

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py:96]: Ah right. The latter is specifically for `Lock`.

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [inline src/trio/_core/_parking_lot.py:274]: Should `ParkingLot.unpark`/`repark` raise `BrokenResourceError` if the lot is broken? Breaking the lot empties it, so `unpark`/`repark` becomes no-ops otherwise.

**Comment by jakkdl** [comment]
> @njsmith cause this is interface design, do you think there's a better way to express all of this?

as we don't actually have anybody requesting the functionality on `ParkingLot` or `add_parking_lot_breaker`/`remove_parking_lot_breaker`, we could opt not to make these public for now and document them as an experimental API. Or I suppose make them public but with a warning text in the docstring that they're experimental.

File: newsfragments/3081.feature.rst
Comment (jakkdl) [inline newsfragments/3081.feature.rst:1]: idk why sphinx isn't finding any of the things I added

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [reply src/trio/_core/_parking_lot.py:274]: I don't think so...

It's hard to think of a concrete case that this matters for; locks will not be able to reach them. However, it's possible to come up with cases I suppose?:
 - case for raising errors: data structure that needs to have at least one task parked at all times. Obviously having none would be a resource error.
 - case against: people putting parks in a try, then unparking the next task in a finally. Throwing would just make the exception worse.

I think the second case convinces me that it shouldn't raise but idk.

File: src/trio/_sync.py
Comment (Zac-HD) [inline src/trio/_sync.py:24]: I'm not convinced we need a new exception type here - wouldn't raising `BrokenResourceError` make sense?  If we do go with a new type, I'd at least inherit from that.

File: src/trio/_core/_parking_lot.py
Comment (Zac-HD) [reply src/trio/_core/_parking_lot.py:274]: yeah, I think treating these as noops makes sense, even though my initial inclination was to error out.

File: newsfragments/3081.feature.rst
Comment (Zac-HD) [reply newsfragments/3081.feature.rst:1]: I think they have to be explicitly documented somewhere in a .rst file, since Trio doesn't use `.. automodule::`.

File: src/trio/_core/_tests/test_parking_lot.py
Comment (Zac-HD) [inline src/trio/_core/_tests/test_parking_lot.py:269]:
```suggestion
    # check that trying to park in broken lot errors
```

File: src/trio/_core/_tests/test_parking_lot.py
Comment (Zac-HD) [inline src/trio/_core/_tests/test_parking_lot.py:235]: add a code comment to explain these duplicated lines?

**Review by Zac-HD — COMMENTED** [review]
This looks good to me!  I'd appreciate a review by another experienced maintainer but I'd be inclined to merge once minor comments are addressed 😁

File: src/trio/_sync.py
Comment (A5rocks) [reply src/trio/_sync.py:24]: Yeah this might have been because of my comment. Sorry for not being more explicit: IMO parking lots should raise a subclass of `BrokenResourceError`. This would mean that locks can do something like this (for better error messages):

```py
try:
  await parking_lot.park()
except BrokenParkingLot as e:
  raise BrokenResourceError(f"this lock was held by {e.task} when it exited")  # note specifically that task is accessible by `e.task`
```

File: src/trio/_core/_tests/test_parking_lot.py
Comment (A5rocks) [reply src/trio/_core/_tests/test_parking_lot.py:235]:
```suggestion
    # check that breaking a parking lot is re-entrant
    trio.lowlevel.add_parking_lot_breaker(task, lot)
    trio.lowlevel.add_parking_lot_breaker(task, lot)
    trio.lowlevel.remove_parking_lot_breaker(task, lot)
    trio.lowlevel.remove_parking_lot_breaker(task, lot)
```
(that's actually a bad explanation maybe, but it's not like test comments matter that much)

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:279]: I'm just now realizing that this relationship is a bit flawed. Lots can only be marked as broken by one task, but multiple tasks could be marked as a parking lot breaker and exit (or a task multiple times, but that's handled already). I'm not sure what's a good idea. We can't raise any error because there's no good place to raise errors, but I suppose we could warn? Or maybe `broken_by` should be a list.

File: newsfragments/3081.feature.rst
Comment (jakkdl) [reply newsfragments/3081.feature.rst:1]: ah yes, this sounds familiar. thanks

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py:279]: should warning be raised if the same task breaks a lot multiple times?

File: src/trio/_core/_tests/test_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_tests/test_parking_lot.py:235]: added comment, slightly different from a5rocks suggestion

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [reply src/trio/_core/_parking_lot.py:279]: Yeah that sounds like a fine warning and a task breaking a lot multiple times is probably fine. After all we allow nesting these add/remove_parking_lot_breaker

File: src/trio/_sync.py
Comment (jakkdl) [reply src/trio/_sync.py]: removed `StalledLockError` and reverted to everything using `BrokenResourceError`, but Lock still reraises. I think that's what the conclusion was?

File: src/trio/_core/_run.py
Comment (Zac-HD) [inline src/trio/_core/_run.py:1866]: IMO we should do this before the misnested-cancel-scopes check up the top; if that happens to fire we still want locks to break instead of hanging.

File: src/trio/_sync.py
Comment (Zac-HD) [inline src/trio/_sync.py:786]:
```suggestion
          BrokenResourceError: if the holder of the underlying lock exits without releasing it.
```

Feel free to ignore, but my personal preference is to say "holder" rather than "owner" of the lock, and in the case of Condition refer to the _underlying lock_ as in the class docstring.

**Review by Zac-HD — APPROVED** [review]
Two minor things below, and then unless @njsmith or @oremanj raise any concerns let's merge after the weekend!

File: src/trio/_sync.py
Comment (jakkdl) [reply src/trio/_sync.py:786]: "owner" is publicly exposed in e.g. `LockStatistics` and already used in several other docstrings, so unless others agree I think it probably makes more sense to be consistent with that.
"underlying" sounds good though

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:99]: These should have a summary.

I think it might be better to link to the `break_lot` to explain the effects of a task exiting without removing themselves.

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:280]: Summary line here too. Also "The error raised contains a reference to the task sent as a parameter." should probably just mention that `broken_by` is available on parking lots, because I personally missed that previously.

File: src/trio/_core/_run.py
Comment (A5rocks) [inline src/trio/_core/_run.py:1826]: Maybe add a line comment here like `# break parking lots associated with a task that exited` because otherwise the warning (from multiple tasks breaking a lot) might point to here in the stack trace.

File: src/trio/_core/_tests/test_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_tests/test_parking_lot.py:313]:
```suggestion
    with RaisesGroup(Matcher(_core.BrokenResourceError, match="^Parking lot broken by")):
```
for consistency

File: src/trio/_sync.py
Comment (A5rocks) [inline src/trio/_sync.py:12]: We could import the functions from `._core` for a nicer import

File: src/trio/_sync.py
Comment (A5rocks) [inline src/trio/_sync.py:602]:
```suggestion
                    f"Owner of this lock exited without releasing: {self._owner}",
```

File: src/trio/_tests/test_sync.py
Comment (A5rocks) [inline src/trio/_tests/test_sync.py:596]: How about testing the full fix because this PR should be the full fix.

**Review by A5rocks — COMMENTED** [review]
Mostly just nitpicks

File: src/trio/_tests/test_sync.py
Comment (jakkdl) [reply src/trio/_tests/test_sync.py:596]: this+next test is the full fix, so the docstring is outdated. Thanks

Although that issue contains some other loose thoughts (give warning if exiting without releasing the lock, semaphore things that got set aside, add this feature to capacitylimiter/semaphore) that should perhaps be mulled over before closing it.

File: src/trio/_sync.py
Comment (jakkdl) [reply src/trio/_sync.py:602]: looking forward to getting the linter rule for this

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py]: linking to `break_lot` sounds good, which removed the need for a summary.

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py]: not super happy with the docstring, but should be fine now? Feel free to suggest changes

File: src/trio/_core/_parking_lot.py
Comment (A5rocks) [inline src/trio/_core/_parking_lot.py:279]: I don't think we do this

File: src/trio/_core/_run.py
Comment (A5rocks) [reply src/trio/_core/_run.py:1903]: Sorry to be specific:
```suggestion
                lot.break_lot(task)  # task exited while it was a parking lot breaker
```
Because this line specifically will be highlighted in the default traceback handler. (even though custom ones can get context)

**Review by A5rocks — APPROVED** [review]
A couple nitpicks but this looks good otherwise!

(I'm not going to lie `add_parking_lot_breaker` was not the best name I've ever come up with. At least it's descriptive...)

File: src/trio/_core/_parking_lot.py
Comment (jakkdl) [reply src/trio/_core/_parking_lot.py:279]: Ah, the sentence is ambiguous as to where that attribute is. fixed

File: src/trio/_core/_run.py
Comment (jakkdl) [reply src/trio/_core/_run.py:1903]: I started playing around with this... and it appears that the `RuntimeWarning` can get transformed into a TrioInternalError >.>

Comment (jakkdl) [reply src/trio/_core/_run.py:1903]: we might also want to raise an error if trying to register a task as a breaker that has already exited

Comment (jakkdl) [reply src/trio/_core/_run.py:1903]: > Nevermind, I was completely incorrect about how warnings seem. I'm concerned that this might cause a weird warning location if someone adds multiple tasks as breakers for a single lot and they all exit, but I'm not sure how to better account for that.

yeah it currently is giving me zero useful location info on where the task exited. And even trying to play around with `stacklevel` and [`skip_file_prefixes` from 3.12](https://docs.python.org/3/library/warnings.html#warnings.warn) I fail to get any change in behavior. Not sure if that's me failing or trio doing sus things

File: src/trio/_core/_tests/test_parking_lot.py
Comment (jakkdl) [inline src/trio/_core/_tests/test_parking_lot.py:272]: @A5rocks ?

File: src/trio/_core/_run.py
Comment (jakkdl) [inline src/trio/_core/_run.py:2806]: This could be too broad (maybe warnings should cause internal errors and prompt people to open issues) or too narrow (if passing through runtimewarnings, why not all warnings?). Could make the `RuntimeWarning` a custom warning class instead

**Review by jakkdl — COMMENTED** [review]
turns out there were some nasty things still buried >.>

File: src/trio/_core/_run.py
Comment (A5rocks) [reply src/trio/_core/_run.py:1903]: At worst we could make the warning into a no-op when in trio internals (just duplicate the checks above this).

... but that's not very useful.

File: src/trio/_core/_tests/test_parking_lot.py
Comment (A5rocks) [reply src/trio/_core/_tests/test_parking_lot.py:272]: I think it's fine, the main issue I think is just warning in trio internals being misleading.

File: src/trio/_core/_run.py
Comment (A5rocks) [reply src/trio/_core/_run.py:1903]: Maybe this is enough justification to make `broken_by` a list? It's a slightly worse API but I think it's fine.

File: src/trio/_core/_tests/test_parking_lot.py
Comment (Zac-HD) [reply src/trio/_core/_tests/test_parking_lot.py:272]: I concur that this seems fine; we do want to give the warning but it's not otherwise a problem - in typical usage you'd get an error about the lot being broken when you go to park in it, and that will likely still happen here (if it hasn't already!).

File: src/trio/_core/_tests/test_parking_lot.py
Comment (Zac-HD) [inline src/trio/_core/_tests/test_parking_lot.py:298]: I think this is only true because we're using `-Werror`, and if we did `with pytest.raises(...)` it'd show up as a warning rather than being raised as an exception.

File: src/trio/_core/_tests/test_parking_lot.py
Comment (Zac-HD) [inline src/trio/_core/_tests/test_parking_lot.py:314]: Yep, [once per location](https://docs.python.org/3/library/warnings.html#warning-filter) is the default, and `with pytest.warns()` clears preexisting filters.

**Review by Zac-HD — APPROVED** [review]
Only action items left AFAICT are to remove the `except RuntimeWarning` block, and optionally to clean up any comments that you want to adjust.  Then let's merge 🚀 

Also, thanks again @jakkdl - this has been a rather long project to fix a relatively small issue, but I think we've ended up with a really nice solution and really appreciate all your work on this 😍

File: src/trio/_core/_run.py
Comment (jakkdl) [reply src/trio/_core/_run.py:2806]: Hmm, I'm torn on this - in large part because `TrioInternalError` prompts users to open an issue to report it, and running pytest with `-Werror` isn't that obscure

Comment (Zac-HD) [reply src/trio/_core/_run.py:2806]: Oh, I see!  Yeah, keeping it with a custom warning type seems good then.

Comment (A5rocks) [reply src/trio/_core/_run.py:2806]: To be honest custom warning type is unnecessary if above `lot.break_lot(task)` we put a `lot.broken_by` check. I'm not sure that indicative of quality of implementation though.

Comment (Zac-HD) [reply src/trio/_core/_run.py:2806]: Requiring that check (and erroring on attempts to break an already-broken lot) seems pretty sensible to me, it's not like `ParkingLot` is an interface that gets called _directly_ very often, and so I don't mind adding some ceremony if that helps catch tricky concurrency bugs.

Comment (jakkdl) [reply src/trio/_core/_run.py:2806]: okay so with the check we no longer raise any warnings on task exit if multiple tasks try to break a task, which means we only ever give a warning if manually breaking a broken lot. That doesn't seem particularly logical to me, and we should probably do one of:
1. make breaking a broken lot a no-op
2. raise an error
3. ...go with `broken_by` being a list (or tuple).

idc much which one though. The warning was added after this convo: https://github.com/python-trio/trio/pull/3081/files/7a1ce5b755b5775153a3525bea793ab7da05fb44#r1764544347 where I argued against the list, but if we no longer want to raise a warning from the internals then maybe the list is the way to go? (that also allows for duplicates in the list).

**Review by Zac-HD — APPROVED** [review]
I like the state we've gotten to, and want to use the feature sometime this year, so I propose merging as-is 🙂

File: src/trio/_core/_run.py
Comment (A5rocks) [reply src/trio/_core/_run.py]: I agree with that ordering (though I might slightly prefer 2 over 1? doesn't matter.)

**Comment by A5rocks** [comment]
> I like the state we've gotten to, and want to use the feature sometime this year, so I propose merging as-is 🙂

We can fix `broken_by`'s representation then merge, everything requires a release anyways :P

**Comment by jakkdl** [comment]
Made `broken_by` a list, and cleaned up `test_parking_lot` that had gotten incredibly messy across all the iterations. I think it's ready to merge at this point after a quick review I didn't do anything dumb at the last second

**Review by A5rocks — APPROVED** [review]
Looks good to me based on a quick review

**Pull request merged**
